### PR TITLE
feat: adds tests and docs for @deepkit/type

### LINKED ISSUE #601: Add test cases and documentation to use @deepkit/type
## Describe what you need | want

Because it seems `class-transformer` is abandoned, we might need an alternative. @deepkit/type is one.

## Do you have already an idea for the implementation?

Yes. Tests and docs already finished. 

Scott

### PR DESCRIPTION
**Pull request by smolinari**
This PR adds tests and some documentation for using @deepkit/type with Typegoose for serialization and deserialization.

## Related Issues

- fixes #601 

Scott

### CODE REVIEW

**Review by hasezoey — CHANGES_REQUESTED** [review]
Thanks for the documentation about `@deepkit/type`
there are some things needing to be fixed in this pr:
- please resolve the merge conflicts (currently `package.json`)
- please remove `package-lock.json` and use yarn
- maybe re-commit with a signed / registered commit (like 9f2160ae55387bf1a3fa7b1c99f66f2706e43b0f)

**Comment by smolinari** [comment]
@hasezoey - I have to ask the stupid question, what is a signed / registered commit? 

Scott

**Comment by hasezoey** [comment]
> I have to ask the stupid question, what is a signed / registered commit?

sorry if i have put it weirdly, i meant for you to either:
- [sign the commit](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work) that is currently there
- or "registering" (changing) your commit (to be) like previously in 9f2160ae55387bf1a3fa7b1c99f66f2706e43b0f

and i ask this because it seems your current commit in this PR is not "linked" to your account (also, to have it accounted into one contributor in the git log)

Note: this is just a suggestion, because i have seen it done previously in the mentioned commit

**Comment by smolinari** [comment]
@hasezoey - I got the GPG key going. Funny thing is, I'm certain I didn't have that when I did the other PR. 

I'm also not sure why there are these conflicts. What am I doing wrong? 🤔 

Scott

**Comment by smolinari** [comment]
I've gotten the commit to be verified now. 

What would you like me to do now though? 😁 

Scott

**Review by hasezoey — CHANGES_REQUESTED** [review]
Please rebase onto the latest upstream `beta` branch (yours is way behind)

**Comment by smolinari** [comment]
I have no clue, if I did it right or not. Very new to rebasing. 🤷 If not, I'll close this PR and start a new one on a fresh fork. 

Scott

**Comment by hasezoey** [comment]
yes it seems like you rebased onto the upstream `beta`, but somehow fc63803706a21b896cb6afacd953277ee38c51de came from that, which basically reverts most of the stuff back again

do you want me to rebase this branch?

**Comment by hasezoey** [comment]
Rebased onto latest upstream `beta`, also dropped the problematic extra commits and squashed the package-lock.json -> yarn into the one before

**Comment by hasezoey** [comment]
also changed commiter info if the first commit (to be the same as the later ones)

**Comment by hasezoey** [comment]
if you have problems updating your local branch, run (while being in the branch of this pr) `git pull -f`

File: test/tests/deepkitType.test.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  let id: string;
+  let accountClassObject: Account;
+  let account: Account;
+  beforeEach(async () => {
+    accountClassObject = new Account();
+    accountClassObject.email = 'somebody@gmail.com';
+    accountClassObject.confidentialProperty = 'secret';
+    account = await AccountModel.create(accountClassObject);
+    // note here that _id is an ObjectId, hence the toString()
+    // otherwise it will have the shape of : { _bsonType: 'ObjectId', id: ArrayBuffer }
```
Comment (hasezoey) [inline test/tests/deepkitType.test.ts:37]: can this comment be removed? (is this residue from copying the class-transformer tests?)

File: test/tests/deepkitType.test.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import { getModelForClass, prop } from '../../src/typegoose';
+
+enum Group {
+  confidential = 'confidential',
+  public = 'public',
+}
+
+class Account {
+  @t.mongoId.group(Group.public)
+  public _id: string;
```
Comment (hasezoey) [inline test/tests/deepkitType.test.ts:11]: are you sure its `string` here? without `@prop` the actual type would be a `mongoose.Types.ObjectId`

File: test/tests/deepkitType.test.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      // groups option includes the property group
+      const access = { groups: [Group.confidential, Group.public] };
+      // deserialize pojo back to an Account instance
+      const deserialized = plainToClass(Account, pojo, access);
+      expect(deserialized).toStrictEqual(accountClassObject);
+      expect(deserialized).toHaveProperty('confidentialProperty');
+    });
+  });
+
+  describe('normal query', () => {
```
Comment (hasezoey) [inline test/tests/deepkitType.test.ts:69]: is it right not to use `accountClassObject` in this block?

File: docs/guides/advanced/deepkit-type.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import { getModelForClass, prop } from '../../src/typegoose';
+
+enum Group {
+  confidential = 'confidential',
+  public = 'public',
+}
+
+class Account {
+  @t.mongoId.group(Group.public)
+  public _id: string;
```
Comment (hasezoey) [inline docs/guides/advanced/deepkit-type.md:23]: are you sure its `string` here? without `@prop` the actual type would be a `mongoose.Types.ObjectId`

(same as in the test)

File: docs/guides/advanced/deepkit-type.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  ```
+
+* or a normal document:
+
+  ```ts
+  // exec returns a Mongoose Object
+  const doc = await AccountModel.findById(id).orFail().exec();
+  const access = { groupsExclude: [Group.confidential] };
+  // serialize Account instance back to a Plain Old Javascript Object
+  const serialized = classToPlain(Account, doc, access);
```
Comment (hasezoey) [inline docs/guides/advanced/deepkit-type.md:61]: is it right to use `classToPlain` in this example (where in for the lean example `plainToClass` is used)?

this does not seem to result in the same as the lean example?

(lean example: POJO -> Class | normal example: Document -> POJO (and not Class))

File: docs/guides/advanced/deepkit-type.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  ```
+
+As you can see from these code examples, there is:
+
+* an `access` or "grouping" definition needed to ouput the required properties for any serializing or deserializing. The `@deepkit/type` author suggests not using groupings at all, as they cause `@deepkit/type`'s performance to deteriorate.
```
Comment (hasezoey) [inline docs/guides/advanced/deepkit-type.md:66]: do you maybe have a link to where this is stated?

File: docs/guides/advanced/deepkit-type.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+* a `Mongodb.ObjectId` decorator (`@t.mongoId`) is built into `@deepkit/type`'s decorators.
+
+The `@deepkit/type`'s author also suggests not putting passwords or other confidential data directly inside entities that have mainly public access. It is better to put them in their own class or classes, so they can be handled without the grouping feature. This allows for the performance of serialization and deserialization to be much better and you can control who sees the confidential data more appropriately i.e. not via serialization steps.
```
Comment (hasezoey) [inline docs/guides/advanced/deepkit-type.md:69]: do you maybe have a link to where this is stated?

File: docs/guides/advanced/deepkit-type.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  public _id: string;
+
+  @t
+  public __v: number;
+
+  @t
+  @prop()
+  public email: string;
+  
+}
```
Comment (hasezoey) [inline docs/guides/advanced/deepkit-type.md]: are you sure no `@t` decorator needs to be executed?

**Review by hasezoey — CHANGES_REQUESTED** [review]
Some things i noticed now that everything is in order and "working"

File: test/tests/deepkitType.test.ts (hunk 1 above)
Comment (smolinari) [reply test/tests/deepkitType.test.ts:37]: Sure, it can be removed. Should I do it or will you?

Scott

File: test/tests/deepkitType.test.ts (hunk 3 above)
Comment (smolinari) [reply test/tests/deepkitType.test.ts:69]: Not completely sure what you mean. The test starting around that line is a serialization. It's showing the doc being serialized to a POJO. 

Scott

File: docs/guides/advanced/deepkit-type.md (hunk 4 above)
Comment (smolinari) [reply docs/guides/advanced/deepkit-type.md:23]: See my comment above. 

Scott

File: test/tests/deepkitType.test.ts (hunk 1 above)
Comment (hasezoey) [reply test/tests/deepkitType.test.ts:37]: you can if you want, if not i will do it later before merge (when other stuff is resolved)

File: test/tests/deepkitType.test.ts (hunk 3 above)
Comment (hasezoey) [reply test/tests/deepkitType.test.ts:69]: with this question i meant why `accountClassObject` was not used in this test block, if it really is intended, could you move the creation of `accountClassObject` to the block at uses it (not the "global" one)?

File: docs/guides/advanced/deepkit-type.md (hunk 4 above)
Comment (hasezoey) [reply docs/guides/advanced/deepkit-type.md:23]: what comment above? do you mean https://github.com/typegoose/typegoose/pull/602#discussion_r701271483, when yes - how does it relate to this being a `string` instead of `ObjectId`?

File: docs/guides/advanced/deepkit-type.md (hunk 5 above)
Comment (smolinari) [reply docs/guides/advanced/deepkit-type.md:61]: Hmm... yeah. Like I said in Discord, the tests don't really show the proper intention of when to use serialization and deserialization of these libraries. In effect, we should be showing an example of a DTO coming in being transformed to a Class to be stored via Typegoose and a Doc coming from Typegoose being transformed to a POJO/ DTO. 

I can rework the docs.

File: docs/guides/advanced/deepkit-type.md (hunk 6 above)
Comment (smolinari) [reply docs/guides/advanced/deepkit-type.md:66]: https://deepkit.io/documentation/type/serialization#groups

File: docs/guides/advanced/deepkit-type.md (hunk 7 above)
Comment (smolinari) [reply docs/guides/advanced/deepkit-type.md:69]: This was stated in a discussion with the author. So not. 

Scott

File: docs/guides/advanced/deepkit-type.md (hunk 8 above)
Comment (smolinari) [reply docs/guides/advanced/deepkit-type.md]: Yes, it is needed, so the properties get de-/serialized.  

Scott

File: docs/guides/advanced/deepkit-type.md (hunk 6 above)
Comment (hasezoey) [reply docs/guides/advanced/deepkit-type.md:66]: thanks, could you maybe add that link to the docs (to provide a source on that statement)?

File: docs/guides/advanced/deepkit-type.md (hunk 8 above)
Comment (hasezoey) [reply docs/guides/advanced/deepkit-type.md]: nevermind i didnt work directly with a decorator for a while

as can be seen in [this line](https://github.com/deepkit/deepkit-framework/blob/be6a253a90a88320f078b8b3ac78d73c0c672e83/packages/type/src/decorators.ts#L1276), the `t` is a alias for `fRaw`, which is defined in [this line](https://github.com/deepkit/deepkit-framework/blob/be6a253a90a88320f078b8b3ac78d73c0c672e83/packages/type/src/decorators.ts#L922), which stores the result of `Field()` (basically already executing it once, so no need for extra `()` when the `t` decorator is used)
and `Field` is [defined here](https://github.com/deepkit/deepkit-framework/blob/be6a253a90a88320f078b8b3ac78d73c0c672e83/packages/type/src/decorators.ts#L850), which uses the same pattern as typegoose does with decorators

File: test/tests/deepkitType.test.ts (hunk 2 above)
Comment (smolinari) [reply test/tests/deepkitType.test.ts:11]: Yes, the @deepkit/type `mongoId` decorator seems to make it work. 

Scott

> also changed commiter info if the first commit (to be the same as the later ones)
> 
> PS: also thanks to the rebase re-commiting the commits they are now signed with my key (you are the author, i am the commiter - [see this stackoverflow question](https://stackoverflow.com/questions/18750808/difference-between-author-and-committer-in-git))

I'm still a GIt noob, so I hope your comment wasn't sarcasm and just an honest thank you. Hehehe.... I really am not sure. 

Scott

File: test/tests/deepkitType.test.ts (hunk 1 above)
Comment (smolinari) [reply test/tests/deepkitType.test.ts:37]: @hasezoey - I think my lack of Git experience is showing. I've got 8 conflicting files now when pulling from my origin and I'm not sure how to resolve them properly. Just force the changes? I don't want to mess any work you did either. So, if you could do the changes, I'd appreciate it.

File: test/tests/deepkitType.test.ts (hunk 3 above)
Comment (smolinari) [reply test/tests/deepkitType.test.ts:69]: Ah. Ok. @hasezoey I see what you mean. `accountClassObject` is used twice though and I did change the initialization code  from `beforeAll` to `beforeEach`, so that object is reinstated. It would mean that code would need to go into the two tests to get back to `beforeAll`.  

Scott

File: docs/guides/advanced/deepkit-type.md (hunk 4 above)
Comment (smolinari) [reply docs/guides/advanced/deepkit-type.md:23]: This comment. https://github.com/typegoose/typegoose/pull/602#discussion_r701597540

Scott

File: docs/guides/advanced/deepkit-type.md (hunk 6 above)
Comment (smolinari) [reply docs/guides/advanced/deepkit-type.md:66]: @hasezoey - As I mentioned above, I'm a noob to Git and resolving conflicts and what to do to not zap your changes. VS-Code is showing me the changes and to select from them. I need to go through 8 different files. I can try this, but with no guarantee of it working properly. 

Is there an easier and safer way to be sure any changes I now make don't screw things up royally? I'll be honest too. I use Gitkraken and the options they give are like chinese to me. 

Pull - Fetch all 
Pull - Fast forward if possible
Pull - Fast forward only
Pull - Rebase

🤷 

I did what was defaulted too, Pull - Fast forward if possible, and now I have these conflicts. 😢 

Scott

File: test/tests/deepkitType.test.ts (hunk 2 above)
Comment (hasezoey) [reply test/tests/deepkitType.test.ts:11]: > I'm still a GIt noob, so I hope your comment wasn't sarcasm and just an honest thank you. Hehehe.... I really am not sure.

it was not really a thank you or sarcasm, i just noted that i changed the commiter information on the original first commit and what the rebase has done to the commits (with a stackoverflow question that explains it really good)

File: test/tests/deepkitType.test.ts (hunk 1 above)
Comment (hasezoey) [reply test/tests/deepkitType.test.ts:37]: i actually added a comment to show how to update your local branch after the rebase i had done: https://github.com/typegoose/typegoose/pull/602#issuecomment-911700895

TL;DR: run `git pull -f` (this will replace the local branch with the changes on the remote by force, meaning it will remove all local changes that do not align with the remote branch)

Comment (hasezoey) [reply test/tests/deepkitType.test.ts]: fixed in 29a667f4aff4dcc8d84bca59c516a72b4de796de

File: test/tests/deepkitType.test.ts (hunk 3 above)
Comment (hasezoey) [reply test/tests/deepkitType.test.ts:67]: > It would mean that code would need to go into the two tests to get back to beforeAll.

no it would not, it would just need to be moved from the `describe("@deepkit/type transforms").beforeEach` to the `describe("@deepkit/type transforms").describe("lean query")` block

but now that i actually looked at it again, no need to move it, but original issue stands, why are the `accountClassObject` / `account` values not used in the `describe("@deepkit/type transforms").describe("normal query")` block?
-> actually as a afterthought, this "why not used here" might be related to also using `classToPlain` in the tests instead of `plainToClass` for `normal query`

File: docs/guides/advanced/deepkit-type.md (hunk 6 above)
Comment (hasezoey) [reply docs/guides/advanced/deepkit-type.md:66]: if that are the only options presented to you (and no Pull - Force), then you will have to go to the directory and manually execute git there, as stated in https://github.com/typegoose/typegoose/pull/602#discussion_r701736086

btw, `Pull - Fast forward` only works when the history did not change between your remote and local branch, but in this case *it did change*, also `Pull - Rebase` would not work, because yes, it would get the remote branch *but* it will try to also apply the old versions of you local branch again ontop of the remote branch - which most probably will lead to conflicts (or at the least, duplicated commits)

**Comment by hasezoey** [comment]
to summarize, the only changes left to do are:
- [ ] add the "groups are slow" statement to the documentation as a reference
- [ ] change the test `normal query` to also use `plainToClass`

File: docs/guides/advanced/deepkit-type.md (hunk 6 above)
Comment (hasezoey) [reply docs/guides/advanced/deepkit-type.md]: added with fa5fa8e258609a5fd8abc43f8ef95afb9e3cb5da

File: test/tests/deepkitType.test.ts (hunk 3 above)
Comment (smolinari) [reply test/tests/deepkitType.test.ts:67]: @hasezoey - I'm lost now. 🤪 

My intention with the tests was to show serialization and deserialization. I needed the `accountClassObject` to compare to the deserialized object, i.e. to the POJO that is transformed via `plainToClass`. 

The groups tests are a bit different than with `class-transformer`. With `@deepkit/type`, the logic is, if a group is added, only that group is de-/serialized and all other properties are excluded. That is why I added the two groups and the explanations in the docs. To me, it isn't a super-duper solution. But, it comes down to performance and keeping grouping ugly is just a matter of trying to suggest avoiding it altogether. 

Scott

**Comment by smolinari** [comment]
> to summarize, the only changes left to do are:
> 
> * [x]  add the "groups are slow" statement to the documentation as a reference ([fa5fa8e](https://github.com/typegoose/typegoose/commit/fa5fa8e258609a5fd8abc43f8ef95afb9e3cb5da))
> * [ ]  change the test `normal query` & documentation to also use `plainToClass`

@hasezoey - I'd like to do this, but like I mentioned above, my repo is borked. What would you suggest I should do to not mess things up even more?

Scott

File: test/tests/deepkitType.test.ts (hunk 3 above)
Comment (hasezoey) [reply test/tests/deepkitType.test.ts:67]: resolved with 1fbe58064faecc415eb50c9f2ea0a27afa5bdc19

File: docs/guides/advanced/deepkit-type.md (hunk 5 above)
Comment (hasezoey) [reply docs/guides/advanced/deepkit-type.md:65]: resolved with 1fbe58064faecc415eb50c9f2ea0a27afa5bdc19

**Review by hasezoey — APPROVED** [review]
everything now looks fine, @smolinari are you fine with the changes and how the documentation looks?

(you can run docusaurus manually, see [website/README](https://github.com/typegoose/typegoose/blob/master/website/README.md))

**Review by hasezoey — APPROVED** [review]
again, everything looks still fine, @smolinari are you fine with the changes and how the documentation looks?

(you can run docusaurus manually, see [website/README](https://github.com/typegoose/typegoose/blob/master/website/README.md))
